pkg/arvo/app/ping.hoon

```diff
@@ -157,24 +288,44 @@
 ::
 ++  on-init
   ^-  [(list card) _this]
-  =^  cards  state  (kick our.bowl now.bowl)
+  =.  plan.state  [%nat ~]
+  =^  cards  state  (kick:ships our.bowl now.bowl)
   [cards this]
 ::
 ++  on-save   !>(state)
 ++  on-load
-  |=  old=vase
-  =.  state  !<(_state old)
-  (on-poke %noun !>(%kick))
+  |=  old-vase=vase
+  |^
+  =/  old  !<(state-any old-vase)
+  =?  old  ?=(%0 -.old)  (state-0-to-1 old)
+  ?>  ?=(%1 -.old)
+  =.  state  old
+  =^  cards  state  (kick:ships our.bowl now.bowl)
+  [cards this]
+  ::
+  +$  state-any  $%(state-0 state-1)
+  +$  state-0  [%0 ships=(map ship [=rift =ship-state])]
+  ::
+  ++  state-0-to-1
+    |=  old=state-0
+    ^-  state-1
+    [%1 ~ 0 %nat ~]
+  --
 ::  +on-poke: positively acknowledge pokes
 ::
 ++  on-poke
   |=  [=mark =vase]
-  ?:  =(q.vase %kick)
-    =.  ships.state
-      %-  ~(run by ships.state)
-      |=  [=rift =ship-state]
-      [999.999 ship-state]
-    on-init
+  ?:  =(q.vase %kick)  :: NB: ames calls this on %born
+    =^  cards  state  (kick:ships our.bowl now.bowl)
+    [cards this]
+  ?:  =(q.vase %nat)
+    =.  plan.state  [%nat ~]
+    =^  cards  state  (kick:ships our.bowl now.bowl)
+    [cards this]
+  ?:  =(q.vase %no-nat)
+    =.  plan.state  [%pub ~]
+    =^  cards  state  (kick:ships our.bowl now.bowl)
+    [cards this]
```

Comment:
thinking if this could be tweaked a bit to avoid the repeating `=^ cards state`

Reply:
I just tried to shrink this function in a few different ways and it was harder than I had expected.  I do think it would be cleaner as a `?+  q.vase`, though.

Reply:
Yeah, I can put the whole thing in =^ like on-arvo and on-agent, which would be more stylistically consistent

Reply:
I think this might work:

```hoon
|=  [=mark =vase]
|^  ?+  q.vase  `this
      %kick    kick  ::  NB: ames calls this on %born
      %nat     kick(plan.state [%nat ~])
      %no-nat  kick(plan.state [%pub ~])
    ==
++  kick
  =^  cards  state  (kick:ships our.bowl now.bowl)
  [cards this]
--
```

Reply:
I tried a few options, and I think I'm going to go with this:
```
  =^  cards  state
    ?:  =(q.vase %kick)  :: NB: ames calls this on %born
      (kick:ships our.bowl now.bowl)
    ?:  =(q.vase %nat)
      =.  plan.state  [%nat ~]
      (kick:ships our.bowl now.bowl)
    ?:  =(q.vase %no-nat)
      =.  plan.state  [%pub ~]
      (kick:ships our.bowl now.bowl)
    `state
  [cards this]
```

Everything with `?+` was harder to read, because it needs either queenside or a `|^`, which is a lot of visual indirection for something this simple.  If you wanted to go further, I think you'd stop using `%noun` mark and make a real one, so that it can be a `?-` and is distinct from the "handling a ping" case.  Then you can just `=.  plan.state` a `?-` and do one `kick` after the whole thing.  But again, I think this is too much indirection for something this simple.

Also realized these should be restricted to local pokes, so I added that check.